modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableView.java

```diff
@@ -2540,68 +2565,40 @@ private void updateTreeEventListener(TreeItem<S> oldRoot, TreeItem<S> newRoot) {
+
+                        List<TreeTablePosition<S, ?>> currentSelection = new ArrayList<>(selectedCellsMap.getSelectedCells());
+                        List<TreeTablePosition<S, ?>> updatedSelection = new ArrayList<>();
 
-                        // (2)
-                        List<TreeTablePosition<S,?>> selectedIndices = new ArrayList<>(getSelectedCells());
-
-                        // (3)
-                        List<TreeTablePosition<S,?>> newIndices = new ArrayList<>(selectedIndices.size());
-
-                        // (4)
                         boolean selectionIndicesChanged = false;
-                        for (int i = 0; i < selectedIndices.size(); i++) {
-                            final TreeTablePosition<S,?> oldIndex = selectedIndices.get(i);
-                            final int oldRow = oldIndex.getRow() - offset;
-
-                            if (pMap.containsKey(oldRow)) {
-                                int newIndex = pMap.get(oldRow) + offset;
-
-                                selectionIndicesChanged = selectionIndicesChanged || newIndex != oldRow;
-
-                                newIndices.add(new TreeTablePosition<>(oldIndex.getTreeTableView(), newIndex, oldIndex.getTableColumn()));
-                            }
-
-                            // check if the root element of this event was selected, so that we can retain it
-                            if (oldIndex.getRow() == startRow) {
-                                newIndices.add(new TreeTablePosition<>(oldIndex.getTreeTableView(), oldIndex.getRow(), oldIndex.getTableColumn()));
+                        for (TreeTablePosition<S, ?> selectedCell : currentSelection) {
+                            int newRow = treeTableView.getRow(selectedCell.getTreeItem());
+                            if (selectedCell.getRow() != newRow) {
+                                selectionIndicesChanged = true;
                             }
+                            updatedSelection.add(new TreeTablePosition<>(selectedCell, newRow));
                         }
-
                         if (selectionIndicesChanged) {
-                            // (5)
-                            quietClearSelection();
-                            stopAtomic();
-
-                            selectedCellsMap.setAll(newIndices);
-
-                            final int offsetOldIndex = oldSelectedIndex - offset;
-                            if (offsetOldIndex >= 0 && offsetOldIndex < getItemCount()) {
-                                int newIndex = e.getChange().getPermutation(offsetOldIndex);
-                                setSelectedIndex(newIndex + offset);
-                                focus(newIndex + offset);
+                            if (treeTableView.isSortingInProgress()) {
+                                startAtomic();
+                                selectedCellsMap.setAll(updatedSelection);
+                                stopAtomic();
```

Comment:
This code block gets executed only as a result of sorting performed by `TreeTableView.sort()` method. The method `TreeTableView.sort()` makes call to the `SelectionModel.startAtomic()` and `SelectionModel.stopAtomic()`. Hence, Ideally it is not required to call those methods from here. But keeping these for precautions to safe guard against any future changes and better understanding when reading.